kernel_tracing/metric_evaluation/test.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2022 Intel Corporation. All rights reserved.
+
+"""
+This script takes as an argument a path to a json file that contains test cases
+These cases specify a bpftrace script to collect data while specified shell commands are run
+When the shell commands finish, the bpftrace script is stopped and variables are collected
+The conditions in the spec file are evaled, replacing @vars with the values from the bpftrace script
+If any condition fails this script exits with code 1, otherwise it exits with code 0
+Shell commands are run in the directory of the spec file
+The bpftrace script path is relative to the spec file as well
+See spec.json for an example
+"""
+
+import os
+import sys
+import json
+import subprocess
+import re
+
+class colors:
+  BLUE = "\033[94m"
+  GREEN = "\033[92m"
+  RED = "\033[91m"
+  END = "\033[0m"
+def print_color(color, s):
+  print(color + s + colors.END)
+
+class BPF:
+  def __init__(self, bpftrace_script):
+    self.script = bpftrace_script
+    self.process = None
+
+  def start(self):
+    self.process = subprocess.Popen(["bpftrace", self.script], stdout=subprocess.PIPE, encoding="utf-8")
+    # Wait for bpftrace script to load
+    while True:
+      line = self.process.stdout.readline()
+      if "Attaching" in line: break
+
+  def stop(self):
+    # Send ctrl-c
+    self.process.send_signal(subprocess.signal.SIGINT)
+    self.process.wait()
+
+    # Collect printed variables for condition evaluation
+    lines = self.process.stdout.read().split("\n")
+    vars = {}
+    for line in lines:
+      if re.match("@.*: .*", line):
+        key, val = line.split(": ", 1)
+        vars[key] = val
+    return vars
+
+def check_condition(condition, vars):
+  for key, val in vars.items():
+    condition = condition.replace(key, val)
+  return eval(condition)
+
+def run_test(test_case, working_dir):
```

Comment:
`check_condition`, `run_test`, `print_color`,... I'm not familiar with `pytest` but a lot of this script looks like re-implementing a test framework. We unfortunately already made that mistake in this sof-test repo + in some internal repos and I'd hate doing it again.

tl;dr: did you take a quick look at `pytest`?

Reply:
The reason I did this all manually is so that the tests can be defined in a JSON file, which makes it a lot simpler to write/manage tests. While I could convert this script to a library and have the tests written in Python, that would reduce the simplicity of writing tests. JSON is also easy to manipulate with a program, so it'd be much easier to design a script to generate test suites that are customized per-device. If you're just as happy to write the test cases in Python, then I'll make this a library that works with `pytest`. Otherwise, I'll look into using `pytest` with the current JSON-based test definitions.

Reply:
> Otherwise, I'll look into using pytest with the current JSON-based test definitions.

Yes.

I'm not very familiar with pytest but I don't see why it would stop you from configuring tests in JSON. I think it's pretty flexible.

Reply:
The issue is that pytest organizes test cases into functions. Therefore, I'd need to create a function for each test case in the JSON which isn't really possible. If each case didn't have it's own function, the error output would be pretty useless as you wouldn't even know which test case is failing. If pytest has some way to dynamically create tests, then it would work great. I'll do some more investigating to see if there is a method for this.

Reply:
I think pytest supports test parameters. So hopefully you can have a single function for all the tests. Random example (found very quickly, there are probably better ones)
https://docs.pytest.org/en/6.2.x/example/parametrize.html

```
  <Function test_timedistance_v2[20011212-20011211-expected0]>
  <Function test_timedistance_v2[20011211-20011212-expected1]>
```

Reply:
After many hours, I was able to wrangle pytest so that it works with the `spec.json`. Getting the output formatting right took forever, but I'm really happy with the end result:
![image](https://user-images.githubusercontent.com/12448243/161403862-7656e2fe-059f-40df-b120-dd92662d142b.png)

Give it a try and let me know if you have any more suggestions.